utilities/src/main/java/org/craftercms/commons/file/stores/S3Utils.java

```diff
@@ -59,8 +57,39 @@ public static final S3Client createClient(AbstractAwsProfile profile, boolean us
         return builder.build();
     }
 
+    /**
+     * Creates an {@link S3AsyncClient} based on the given profile config.
+     *
+     * @param profile the configuration profile
+     *
+     * @return a client to an S3AsyncClient account
+     */
+    public static final S3AsyncClient createAsyncClient(AbstractAwsProfile profile, boolean useCustomEndpoint) {
+        S3AsyncClientBuilder builder = S3AsyncClient.builder()
+                .credentialsProvider(profile.getCredentialsProvider());
+
+        if (useCustomEndpoint && StringUtils.isNotEmpty(profile.getEndpoint()) &&
+                StringUtils.isNotEmpty(profile.getRegion())) {
+            builder.endpointOverride(URI.create(profile.getEndpoint()))
+                    .region(Region.of(profile.getRegion()));
+            if (profile instanceof S3Profile) {
+                builder.serviceConfiguration(S3Configuration.builder()
+                        .pathStyleAccessEnabled(((S3Profile) profile).isPathStyleAccessEnabled())
+                        .build());
+            }
+        } else if (StringUtils.isNotEmpty(profile.getRegion())) {
+            builder.region(Region.of(profile.getRegion()));
+        }
+
+        return builder.build();
+    }
+
     public static final S3Client createClient(AbstractAwsProfile profile) {
         return createClient(profile, true);
     }
 
+    public static final S3AsyncClient createAsyncClient(AbstractAwsProfile profile) {
+        return createAsyncClient(profile, true);
```

Comment:
How do we avoid leaking these clients?
Do we need to/shoud we pool them?

Reply:
https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/best-practices.html#bestpractice1
We should be able to reuse it. Otherwise, the consumers of these methods should be calling `close()` on the client instance when no longer needed.

Reply:
These clients are created by factory classes which extends `AbstractAwsClientCachingFactory` so they are reuseable and will be closed by the cache.
https://github.com/craftercms/commons/blob/develop/utilities/src/main/java/org/craftercms/commons/aws/AbstractAwsClientCachingFactory.java#L46

```
Factory class that creates AWS clients based on AbstractAwsProfiles. It also caches the clients so instead of creating a new client for the same profile, it will reuse a previous client. Clients that haven't been used for a while will be evicted from the cache and shutdown.
```